### DISCUSSION TITLE
How to mirror path for Red

### DISCUSSION THREAD

**Discussion by Ajeff22209** [mjansen4857/pathplanner#626] [General]
I've been reading the docs and it isn't really clear how to mirror an auto for the red alliance. I'm using CTRE's example code that integrates pathPlanner with the CTRE swerve library and it configures the auto builder here: 

 ```
 AutoBuilder.configureHolonomic(
            ()->this.getState().Pose, // Supplier of current robot pose
            this::seedFieldRelative,  // Consumer for seeding pose against auto
            this::getCurrentRobotChassisSpeeds,
            (speeds)->this.setControl(autoRequest.withSpeeds(speeds)), // Consumer of ChassisSpeeds to drive the robot
            new HolonomicPathFollowerConfig(new PIDConstants(10, 0, 0),
                                            new PIDConstants(10, 0, 0),
                                            TunerConstants.kSpeedAt12VoltsMps,
                                            driveBaseRadius,
                                            new ReplanningConfig()),
            ()->false, // Change this if the path needs to be flipped on red vs blue
            this); // Subsystem for requirements
```
            
            I tried to flip it by changing ()->false to 

````
           ()->{
                var alliance = DriverStation.getAlliance();
                if (alliance.isPresent()) {
                    return alliance.get() == DriverStation.Alliance.Red;
                }
                return false;
            }
````

but that didn't seem to work. the docs say that

> The AutoBuilder configuration requires a method that will return true when a path should be flipped to the red side of the field. The origin of the field coordinate system will remain on the blue side.

> If you wish to have any other alliance color based transformation, you must implement it yourself by changing the data passed to, and received from, PathPlannerLib's path following commands.

It's not really clear how I would implement this, can someone help me? [The repo can be found here](https://github.com/Pantherbotics/FRC-2024-Crescendo). I apologize if I am missing something obvious, it's my first year working with java, and unfortunately I'm the only programmer on my small team

**Comment by WarrenReynolds** [comment]
That bit of code handles running red auto paths that have been designed on path planners blue side.  All autos should be constructed on the blue side of the field.  The problems you have with a fixed original at the blue corner are, Joystick X and Y axis need to be inversed when you are on the red alliance.   (NB: I'm talking holonomic robots with field centric driving here, ie when you push the joystick away from yourself the robot moves away from you regardless of the direction it is facing.)
The reason you have a fixed origin at the blue corner is the april tag co-ordinate are all based off blue being 0,0.  

Other issue that you might run into with blue always being 0,0 are, say if you had a turreted under floor intake (like us) and you had a command that was called loadIntakeFromRobotRightSide, this might work perfectly when you are running a blue auto, but when you are on the red alliance and running the same auto, the robot will be moving left towards the note instead of right because the paths are flipped, due to this code.

```
           ()->{
                var alliance = DriverStation.getAlliance();
                if (alliance.isPresent()) {
                    return alliance.get() == DriverStation.Alliance.Red;
                }
                return false;
            }
```
So inside your auto commands you have to check for Alliance colour, because anything that is side based will need to be inverted when running red path planer auto from the red alliance.

In regards to how to implement it, just use the same code structure as above.
Get the alliance colour, if it exists (not sure why we would need to check this but apparently you have to), then you can check if its red, if it is flip any logic in your command that are side based.

Hope this helps.
Cheers
Warren
